### DISCUSSION TITLE
Sketch execution during the captive portal loop

### DISCUSSION THREAD

**Discussion by Koxx3** [Hieromon/AutoConnect#326] [Q&A]
Hello,

i've seen the explanation to keep the sketch running while waiting for wifi connection.
(cf last chapter here : https://hieromon.github.io/AutoConnect/adcpcontrol.html#disable-the-captive-portal)
This solution is complicated to use to keep the main loop running on big projects with real time constrains.
Is there any other way like running the portal loop in a FreeRTOs task ? if yes, what are the difficulties ?

**Comment by Koxx3** [comment]
here is my current implementation :
https://github.com/Koxx3/SmartController_SmartDisplay_ESP32/blob/master/src/WifiSettingsPortal/WifiSettingsPortal.cpp

To be clear, I'd like my main loop running whatever the autoconnect/portal state ;)

**Comment by Hieromon** [comment]
I would like to ask you about the basic strategy it has for the code you have presented.
Your AUX page contains JQuery, but if the ESP module doesn't have a connection with any AP, there is no internet transparency. So, will the AUX page work properly in the captive portal state?

And I don't quite understand what you mean.
> This solution is complicated to use to keep the main loop running on big projects with real time constrains.
Is there any other way like running the portal loop in a FreeRTOs task ? if yes, what are the difficulties ?

Do you want to execute a captive portal loop by AutoConnect and sketch-specific processing concurrently?

**NOTE**
This topic has converted to the [discussions](https://github.com/Hieromon/AutoConnect/discussions).

**Reply by Koxx3** [reply]
ok, first, for my strategy.
Yes, for updates, I use 3 ways which lead to the same goal : connect the ESP to an access point and launch and OTA update from github firmware storage.
The 3 ways : 
- the user connect the ESP to an access point with AutoConnect and load the flashing page. The JS code will execute (and hide specific page elements which should be displayed only if internet is not available). He selects the release and submit the release ID to the ESP. 
- the user connect the ESP only with his smartphone (no access point configured in autoconnect) but has configured an OTA access point somewhere is in the settings. The Autoconnect page load without internet access (android isn't able to access internet even if 3G/4G is still active). The user has to manually enter a release version ID and submit it. Then the ESP will connect to the hotspot and OTA the specific release.
- the user use android specific application with bluetooth communication (which list versions and submit the firmware ID to the ESP). ESP switch from BLE to wifi access point and flash.

This is not perfect, but It cover most use cases I imaganed.

Why not in access point all the time ? I want to use Autoconnect as the way to access a dynamic dashboard even when riding the escooter. The ESP run a lot of tasks continuously. So ... I need the Autoconnect in CaptivePortal mode (to display the dashboard, settings and co) AND running the main loop for everything else (ADC / DAC / TFT screen / multiple UARTs ...).
So the answer to your last question is : YES ;) 

I placed the portal loop in a FreeRTOs tasks but it seems to crash sometimes (without exception to decode... arghhhh).

**Reply by Hieromon** [reply]
If I understand correctly, the following obstacles are hiding behind your request:

- The TCP stack on the ESP32 Arduino core is not asynchronous. Even if you run it in a separate thread, it will probably destroy memory if the upper layers do not support it.
- It is the client device that creates the captive portal state. When the OS (WiFi connection service) of the client device enters the captive portal state, the web browser is automatically started. Not limited to AutoConnect, WebServer class handlers on the ESP32 cannot break out of the loop created by the client device OS to establish a WiFi connection.
- Only `xTaskCreatePinnedToCore` is possible to achieve multithreading using the ESP32 Arduino framework. It doesn't work with `xTaskCreate`.

To quote the [ESP32 Wi-Fi Programming Model](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/wifi.html#esp32-wi-fi-programming-model) from the ESP-IDF document, we can see that it is only event-driven. The TCP stack needs to work asynchronously and the handlers need to be multiplexed to meet your needs.
<img src="https://docs.espressif.com/projects/esp-idf/en/latest/esp32/_images/blockdiag-3e9661277e869fa79ee746e67c21966774cbfb1d.png">

Therefore, it is necessary to run asynchronous TCP and asynchronous WebServer on the thread created by xTaskCreatePinnedToCore. However, AutoConnect does not support [AsyncWebServer](https://github.com/me-no-dev/ESPAsyncWebServer).
So, I think that it cannot meet your needs.

**Reply by Koxx3** [reply]
understood. It seems to work with the xTaskCreatePinnedToCore

 ```
 xTaskCreatePinnedToCore(
      taskProcessWifiBlocking,   // Function that should be called
      "taskProcessWifiBlocking", // Name of the task (for debugging)
      10000,                     // Stack size (bytes)
      NULL,                      // Parameter to pass
      0,                         // Task priority
      NULL,                      // Task handle,
      1);                        // Core
```
